File: ashok/gate questions/python/ec2010_39.py
# Define logic gate functions
def not_gate(x):
    return int(not x)

def and_gate(x, y):
    return x & y

# Define the 4x1 MUX logic based on inputs A, B and inputs C, D
def mux_4x1(A, B, C, D):
    # I0 = C
    # I1 = D
    # I2 = NOT C
    # I3 = C AND D
    I0 = C
    I1 = D
    I2 = not_gate(C)
    I3 = and_gate(not_gate(C), not_gate(D))
    
    # Select line value (0 to 3)
    select = (A << 1) | B

    # MUX output
    inputs = [I0, I1, I2, I3]
    return inputs[select]

# Collect and display minterms where output is 1
minterms = []

print("A B C D => F")
for A in [0, 1]:
    for B in [0, 1]:
        for C in [0, 1]:
            for D in [0, 1]:
                F = mux_4x1(A, B, C, D)
                # Calculate the minterm number as ABCD binary to decimal
                index = (A << 3) | (B << 2) | (C << 1) | D
                print(f"{A} {B} {C} {D} => {F}")
                if F == 1:
                    minterms.append(index)

# Print final result
print("\nMinterms where F = 1:")
print(f"F = Σm{tuple(minterms)}")

File: ashok/gate questions/c/EC2010_39.c
#include <stdio.h>

// Logic gate functions
int not_gate(int x) {
    return !x;
}

int and_gate(int x, int y) {
    return x & y;
}

// 4x1 MUX logic based on A, B select lines and C, D inputs
int mux_4x1(int A, int B, int C, int D) {
    int I0 = C;
    int I1 = D;
    int I2 = not_gate(C);
    int I3 = and_gate(not_gate(C), not_gate(D));

    int select = (A << 1) | B;

    int inputs[4] = {I0, I1, I2, I3};

    return inputs[select];
}

int main() {
    int minterms[16];
    int count = 0;

    printf("A B C D => F\n");

    for (int A = 0; A <= 1; A++) {
        for (int B = 0; B <= 1; B++) {
            for (int C = 0; C <= 1; C++) {
                for (int D = 0; D <= 1; D++) {
                    int F = mux_4x1(A, B, C, D);
                    int index = (A << 3) | (B << 2) | (C << 1) | D;
                    printf("%d %d %d %d => %d\n", A, B, C, D, F);
                    if (F == 1) {
                        minterms[count++] = index;
                    }
                }
            }
        }
    }

    // Print minterms
    printf("\nMinterms where F = 1:\nF = Σm(");
    for (int i = 0; i < count; i++) {
        printf("%d", minterms[i]);
        if (i < count - 1) {
            printf(", ");
        }
    }
    printf(")\n");

    return 0;
}

